### PR TITLE
merge JSoar compatibility branch

### PR DESCRIPTION
**Pull request by garfieldnate**
I've included all of the commits from the jsoar_compatibility branch, doing a bit of squashing and editing here and there. I have some open questions for @marinier, though, and I don't want to merge this just yet:

1. The lexer/parser have kind of been hacked to support falling back to `exec` when a native RHS function isn't found. Does JSoar have the same behavior (not the hackiness but the fallback)? I didn't see it here: https://github.com/soartech/jsoar/blob/maven/jsoar-core/src/main/java/org/jsoar/kernel/parser/original/OriginalParserImpl.java#L1281
2. The changes to `TclClientMessageEventCallback` require some commenting. It uses a `static` variable, so the assumption here is that the effect of the function will only be allowed to happen exactly once during the entire lifetime of the program. Why is this assumption valid? And should this be applied for the other SML client languages?
3. Would be nice to have tests that exercise the new set math RHS functions, as well as `from-st-xml`.
4. a43477d647767dbd5bfd52bdf2d7e5cf61f37396 leaves behind some commented code. If we're sure of the solution (`thread.detach()` instead of `thread.join()`), then I'd just remove the commented code (and leave the nice comments about why this is being done). On the other hand, the comment claims that this leads to some memory leaks, so I'm not sure if this is right; there's a lot of other dead thread cleanup code, too. Why does `lib_thread` (`launch_tcl`) finish (comment says it's "dead" already) without finishing all the cleanup at the end of `launch_tcl`? Is there an exception being thrown that we miss somewhere?
5. The JSON stuff isn't finished; when you tell a JSON element to generate a string, it instead generates XML. The JSON codebase was copied from the XML codebase, and Norb wrote that it wasn't finished. I don't quite know what the plan for it was. For example, the lines commented out with `NORB` in `rhs_functions.cpp`. Does the `write` RHS function normally write out XML? And was Norb trying to change it to output JSON instead? Sounds nice, actually, but it's not done and I don't have time to finish it now. My plan is to just remove these commits altogether, and if the topic comes up again then we'll be able to revisit the jsoar_compatibility branch. There's a compilation issue on Windows for the unit tests, anyway. Does this sound okay to you?

### CODE REVIEW

**Comment by marinier** [comment]
1. The `exec` RHS function is actually a hack in CSoar that was necessary because SML does not provide a mechanism for the user to register RHS functions directly with the agent (like the built-in functions do). At the time we designed this, it was because we were thinking about the remote application case, and how would you transport a RHS function to be registered over the network? In retrospect, I think we put way to much emphasis on the remote case, and in practice it's really only used by the debugger. And no one would register a remote RHS function because it's SO SLOW. Anyway, JSoar doesn't have this restriction, and so `exec` was never necessary there. Hence, a lot of existing SoarTech code that was originally written for JSoar, but which we sometimes run in CSoar, does not use `exec` to call RHS functions -- they are just called directly exactly like the built-in functions. The change was made in CSoar to support these jsoar agents (that is, to make CSoar more jsoar compatible :)). I think this capability needs to stay. There may be a better implementation, but if you take this out then ST will be forced to continue maintaining its own branch.

2. I'll have to look into this, not sure off the top of my head.

3. The XML stuff has tests in jsoar that we can port: https://github.com/soartech/jsoar/tree/maven/jsoar-core/src/test/java/org/jsoar/kernel/io/xml

4. IIRC, we were having the issue where the Soar Java Debugger would often hang on shutdown if Tcl was loaded. (It may also have been necessary to actually do something with Tcl, not just load it). Norb was attempting to fix that issue, and IIRC this DOES fix the issue in that it no longer hangs. I don't know what the memory leak is about, but my guess is that since it was happening during shutdown, it didn't seem like a priority to fix (and as you can tell from all the commented code, no one really understood what was going on and it was a lot of trial and error). Unfortunately, Norb left ST years ago, so if we wanted to ask him we'd have to track him down, but I'm guessing he wouldn't remember anyway.

5. The JSON stuff can be removed. There is json support in jsoar, but it's rarely used and clearly we never relied on it in CSoar.

**Comment by garfieldnate** [comment]
@marinier Thank you for your helpful comments!

3. We do already have a file with a giant list of JSoar unit tests to backport, so I'll add `SoarTechXmlToWmeTest` to the list. I noticed also that JSoar has multiple `from-*-xml` RHS functions: `from-xml`, `from-at-xml`, `from-st-xml`, `from-mt-xml` and `from-xml`. This branch only implements `from-st-xml`. Is this a problem? If there's a document somewhere stating the differences between JSoar and CSoar, it might be good to note this there.

**Comment by garfieldnate** [comment]
I've pushed some docs updates and removed the JSON support. After comments on 2 and 3 above, I'll merge the branch. I'll add notes from our discussion to the merge commit message.

**Comment by marinier** [comment]
Regarding the changes to `TclClientMessageEventCallback`, this is actually one of the breaking changes that needs to be documented. The issue has to do with how user-defined RHS functions return results. SML only supports returning strings from custom RHS functions. Originally, in order to keep a C interface, we assumed a fixed maximum length for that result. This caused problems, so we instead returned a `std::string` to support variable length results without having to specify a maximum. But of course now the API was not C-only. Practically this meant that to use Soar (especially on linux) you had to recompile for your machine from source. Precompiled binaries usually didn't work because there wasn't a pure C interface.

So Norb modified this to make it a pure C API that supports dynamically creating a buffer of the needed size. I have to look closer at the details, but IIRC the caller provides a char* buffer, and if it's not big enough, the result indicates how big it needs to be (via the `bufSize` output param) and the caller can make a new buffer of the correct size to get the result. The result is cached, so you're not running the RHS function twice. For Tcl (and this would apply to all the other languages) this buffer handling is done in the language handling code, so it's completely transparent to end users. But for anyone who is using SML from C/C++, they would have to modify their functions to handle this. The modification is easy, but needs to be documented.

It looks like the interface files / manual code for Java, Python, and C# has also been updated already to handle this. It also looks like this change isn't limited to just RHS functions but covers other callbacks as well.

Details:

In `TclClientMessageEventCallback` line 199, it checks if the result from the Tcl RHS function will fit in the provided buffer. If not, it changes `bufSize` (which is an output param) to the necessary size, caches the result, and returns NULL. The caller can then see that they got a NULL result and call again with a buffer of the provided size. This time, line 169 sees that we have a cached result, copies it to the provided buffer (which should be big enough now), clears the cache, and returns the buffer.

**Comment by marinier** [comment]
Regarding the different versions of XML support in jsoar, we only use the one that was ported to CSoar. The others are extremely niche.

I don't know if there is a comprehensive document of the differences between CSoar and JSoar, but I think it could be summarized as:
* The core functionality (e.g., the rete, decision cycle, etc.) was ported directly from CSoar and should be functionally identical at the time it was ported or updated.
* The more recent versions of chunking, epmem, smem, LTIs, and activation have not been ported -- jsoar has older versions of these.
* JSoar does not support SVS. (I'd like to, just never got funding to do it.)
* JSoar's I/O is completely different (and I think much nicer). There are actually lots of ways to do I/O in JSoar depending on your needs (low-level high performance and high-level lower performance). They can be mixed and matched. There is no SML support.
* JSoar supports inline javascript, python, clojure, and maybe some other stuff. E.g. you can write a javascript RHS function or I/O handling directly in your Soar file. The support per language varies. Javascript is the best supported.
* JSoar supports Tcl, but it's using jtcl which hasn't been updated in years and is stuck somewhere between tcl 8.4 and 8.5. We use Tcl extensively in much of our code.
* Because there's no SML support, you can directly register RHS functions with an agent just like the built-in functions. E.g., the results are not restricted to being strings.
* JSoar supports remote debugging through it's legilimens module, which basically creates a webserver you can connect to with a simple debugging interface you can open in a browser. But practically we don't use this and instead find ways to run the real debugger when needed. Other than this, JSoar does not natively support remote connections, but specific projects have written various networking support that they needed.
* JSoar does not have a concept of a "kernel" or a "scheduler" (which are part of SML in CSoar). Typically we run each agent in a separate thread (using `ThreadedAgent`) and let them run independently. We have never had a real world use case where running multiple agents in the same thread and having them take turns made sense. I think CSoar has been led astray here by the toy problems it is typically used on (like Eaters and TankSoar), and that it really should throw away the kernel and scheduler and support running agents in separate threads like JSoar.

There's lots of other stuff, too. JSoar has a fairly extensive wiki that should be up-to-date that covers some of it's specific features like `ThreadedAgents`, various I/O mechanisms, and support for embedded languages. https://github.com/soartech/jsoar/wiki

**Comment by garfieldnate** [comment]
>This time, line 169 sees that we have a cached result, copies it to the provided buffer (which should be big enough now), clears the cache, and returns the buffer.

I think maybe I don't have a clear idea of the usage of this thing. This is called by Soar when a Tcl/Python/Whatever function is `exec`'d from an RHS, right? Then it would seem that Soar itself needs to be checking the bufSize value somewhere, but I can't see where this is happening. Can you point it out to me? You said it's happening in manual/interface files for Python, Java and C#.

Also, this means that Soar must absolutely be single-threaded, right? 

```c++
		if ( !prevResult.empty() )
		{
			strncpy( buf, prevResult.c_str(), *bufSize );
			
			prevResult = "";

			return buf;
		}
```

Because this logic absolutely would break if another thread came in and called an RHS function before the first thread could make the second call with the larger buffer.

Also, this cacheing is only implemented for Tcl, but shouldn't it also be implemented for all of the supported languages? Because if a client registers a function which is not idempotent or has side-effects or is expensive, having a function called multiple times like this could be a nasty surprise.

* not cached in [CSharp](https://github.com/SoarGroup/Soar/blob/961_jsoar_compat_rebase_manual_2/Core/ClientSMLSWIG/CSharp/CSharpCallbackByHand.h#L833)
* not cached in [Java](https://github.com/SoarGroup/Soar/blob/961_jsoar_compat_rebase_manual_2/Core/ClientSMLSWIG/Java/JavaCallbackByHand.h#L1111)
* not cached in [Python](https://github.com/SoarGroup/Soar/blob/961_jsoar_compat_rebase_manual_2/Core/ClientSMLSWIG/Python/Python_sml_ClientInterface.i#L283)

**Comment by marinier** [comment]
Sorry, you're right, the CSharp, Java, and Python stuff needs to be updated. I saw they had part of the solution (returning null if the buffer is too small), but clearly looked too fast to see the rest was missing.

Regarding thread safety, Soar is absolutely single threaded right now, so thread safety was not a consideration in the design. That said, I agree these static variables would need to be turned into agent-specific variables to support threading. Each agent should have its own instance of the interpreter, so there should be a path forward here. Also, I'm pretty sure that Tcl and Python only support single-threaded access anyway, but if each agent has its own instance, then only that agent's thread would be accessing it anyway. Note that in JSoar it's important that all interaction with a `ThreadedAgent` happens on that agent's thread, including things like loading productions, registering RHS functions, and registering for events. If you try to do these things from another thread (e.g., the main application thread), then bad things may happen (although if the agents aren't running when you do this you may get away with it). I assume threaded agents in CSoar would need to deal with similar issues.

Regarding how all this works, you need to look at `Kernel::ReceivedRhsEvent` in `sml_ClientKernel.cpp` starting at line 701. This is what actually calls the user RHS function. Since the user does not have direct access to Soar itself, SML registers its own RHS function (`exec`) and this is the handler for that. This in turn calls the user's RHS function. It creates a 1024 buffer, calls the function, and if it gets back NULL increases the buffer to the necessary size and calls it again. If the RHS function properly implements the caching scheme (as demonstrated in the Tcl-specific code) then this all works properly. This is what needs to be documented -- how to properly create your RHS function so this all works.

For Tcl, there is only one RHS function (`tcl`) which similarly delegates to whatever the user-defined RHS functions are. Thus, we are able to have just one RHS function (from SML's perspective) that does the caching, and someone writing Tcl RHS functions doesn't need to worry about that -- it's handled for them. Python at least should be the same -- the C interface is surprisingly similar to Tcl. I'm less certain about how the Java and CSharp interfaces work, but hopefully they can also hide the details from the user. Hopefully a user only needs to worry about it when writing their own C++ RHS functions, and even then the pattern is very simple (it can be copy/pasted). Perhaps we can figure out how to hide this from the user even at this level?

**Comment by garfieldnate** [comment]
Thank you for the detailed comments! Regarding documentation, where should that be? Is there a user guide for writing C++ RHS functions anywhere? I think if it's just in a code comment somewhere it'll be ignored. We do have some Doxygen configs but I haven't actually seen these built.

**Comment by marinier** [comment]
I was thinking of a couple places:
* The manual section on User RHS functions (bottom of page 80 in the version I'm looking at) already shows a tiny bit of C++, so we could expand on that.
* This page has SML examples that would need to be updated with explanations: https://soar.eecs.umich.edu/articles/articles/soar-markup-language-sml/78-sml-quick-start-guide
* And of course the comments in the code itself should be updated, ideally including the specific code that should be added to user RHS functions.

**Comment by marinier** [comment]
BTW, I hope the manual will be distributed with Soar again. I think it was a big mistake to stop doing that. Nothing prevents more recent versions from being posted somewhere, but the distribution should definitely have one.

**Comment by garfieldnate** [comment]
I'm hoping to do a proper release with everything included, so I'll keep that in mind. I see it's there in the [release-support](https://github.com/SoarGroup/Release-Support) repository.

**Comment by marinier** [comment]
I'm recalling another change that should be documented. Currently in CSoar, if a rule referenced a RHS function that doesn't exist, that rule would fail to load. At ST we have some agents that get reused across projects, and some of those agents refer to RHS functions that don't exist in some projects, but that's ok because those rules will never fire in those cases.

In order to support loading rules like this, we changed the behavior to me more like jsoar, where the rules will load, and a runtime error is printed if the rule fires. (If the rule has multiple effects, the other effects still work, like changes to working memory, but the RHS function call will fail with an error). In jsoar, a warning will actually get printed, but the csoar version just says "adding exec to RHS function: <func>", which could be a clue but isn't really the same thing. I think the issue is that the place that checks this information doesn't have access to SML's registered RHS functions, so it's hard to do better.

Probably this should be documented, too. I'm not sure that the current behavior is actually documented anywhere, so maybe this just needs to go in the release notes.

**Comment by marinier** [comment]
I'm recalling another change to be aware of. Currently in CSoar, if a rule referenced a RHS function that doesn't exist, that rule would fail to load. At ST we have some agents that get reused across projects, and some of those agents refer to RHS functions that don't exist in some projects, but that's ok because those rules will never fire in those cases.

In order to support loading rules like this, we changed the behavior to me more like jsoar, where the rules will load, and a runtime error is printed if the rule fires. (If the rule has multiple effects, the other effects still work, like changes to working memory, but the RHS function call will fail with an error). In jsoar, a warning will actually get printed, but the csoar version just says "adding exec to RHS function: <func>", which could be a clue but isn't really the same thing. I think the issue is that the place that checks this information doesn't have access to SML's registered RHS functions, so it's hard to do better.

Probably this should be documented, too. I'm not sure that the current behavior is actually documented anywhere, so maybe this just needs to go in the release notes.

**Comment by garfieldnate** [comment]
@marinier One more question that's bugging me: why did the file have to be C and not C++ in order for the built files to be portable to other systems? What files in Soar have this constraint? I think this is an important maintenance issue that should be documented somewhere.

**Comment by marinier** [comment]
Here's some info on why you need to export C functions. It basically boils down to C++ not having a standard ABI:

See the "Interoperability of Libraries" section at the end: https://gernotklingler.com/blog/creating-using-shared-libraries-different-compilers-different-operating-systems/

https://stackoverflow.com/questions/67839008/please-explain-the-c-abi
